Max concurrent streams

kunalekawde opened the pull request:
#3806 referring this here as in 3806 git local branch was not created leading to confusion.

jay commented on docs/libcurl/opts/CURLMOPT_MAX_CONCURRENT_STREAMS.3:
```diff
@@ -0,0 +1,54 @@
+support.
+Valid values range from 1 to 2147483647 (2^31 - 1) and defaults to 100.
+The value passed here would be honoured based on other system resources properties,
+by which maximum concurrent streams could be achievied. So, its not a guarantee to
```
achievied/achieved  its/ it's but anyway this paragraph is confusing I'd nix it

jay commented on docs/libcurl/opts/CURLMOPT_MAX_CONCURRENT_STREAMS.3:
```diff
@@ -0,0 +1,54 @@
+.\" *
+.\" **************************************************************************
+.\"
+.TH CURLMOPT_MAX_CONCURRENT_STREAMS 3 "08 Aug 2019" "libcurl 7.66.0" "curl_multi_setopt options"
```
7.67 November 6

jay commented on docs/libcurl/opts/CURLMOPT_MAX_CONCURRENT_STREAMS.3:
```diff
@@ -0,0 +1,54 @@
+curl_multi_setopt(m, CURLMOPT_MAX_CONCURRENT_STREAMS, 200L);
+.fi
+.SH AVAILABILITY
+Added in 7.64.3
```
also 7.67

jay commented on docs/libcurl/symbols-in-versions:
```diff
@@ -319,6 +319,7 @@ CURLMOPT_MAXCONNECTS            7.16.3
 CURLMOPT_MAX_HOST_CONNECTIONS   7.30.0
 CURLMOPT_MAX_PIPELINE_LENGTH    7.30.0
 CURLMOPT_MAX_TOTAL_CONNECTIONS  7.30.0
+CURLMOPT_MAX_CONCURRENT_STREAMS  7.66.0
```
7.67

jay commented on include/curl/multi.h:
```diff
@@ -448,6 +451,9 @@ typedef int (*curl_push_callback)(CURL *parent,
                                   void *userp);
 
+/* value for MAXIMUM CONCURRENT STREAMS upper limit */
+#define INITIAL_MAX_CONCURRENT_STREAMS ((1U << 31) - 1)
```
why not int-max

kunalekawde replied:
Yeah, I kept this inline to define done at nghttp2, so that it remains line with exact number considering 32 bit arch.

jay commented on lib/multi.c:
```diff
@@ -3210,3 +3220,10 @@ void Curl_multi_dump(struct Curl_multi *multi)
 #endif
+
+size_t Curl_multi_max_concurrent_streams
+(struct Curl_multi *multi)
```
make this one line

jay commented on lib/multihandle.h:
```diff
@@ -133,6 +133,7 @@ struct Curl_multi {
   struct curltime timer_lastcall; /* the fixed time for the timeout for the
                                     previous callback */
   bool in_callback;            /* true while executing a callback */
+  long max_concurrent_streams; /*max concurrent streams client to support*/
```
add space /*

kunalekawde commented:
> 
> 
> apologies for the flurry of individual comments i tried pushing the rest as a single notification

Not an issue, time spent on review is more appreciated, thanks.

jay commented:
I tried to resolve one comment and somehow resolved all of them. Unwanted magic. The int_max one is resolved the others need to be addressed.

kunalekawde commented:
@jay I've updated commit with resolved comments, could you please check once, let me know if any other steps are required to take this further. 
On the build failure:
1) For Travis: 
    gpg: requesting key BA9EF27F from hkp server keyserver.ubuntu.com
    gpg: keyserver timed out
    gpg: keyserver receive failed: keyserver error
2) For Appveyor:
    a. The RPC server is unavailable
    b. Not sure if below is actually some error.
          === Start of file trace2050
          08:57:09.093000 == Info: STATE: INIT => CONNECT handle 0x1def4f093a8; line 1491 (connection #-5000)
         08:57:09.093000 == Info: Connecting to hostname: connect.example.com.2050
         08:57:09.093000 == Info: Connecting to port: 8990
         08:57:09.093000 == Info: Added connection 0. The cache now contains 1 members
         08:57:09.093000 == Info:   Trying 127.0.0.1:9013...
         08:57:09.093000 == Info: TCP_NODELAY set
         08:57:09.093000 == Info: STATE: CONNECT => WAITCONNECT handle 0x1def4f093a8; line 1547 (connection #0)
        08:57:11.328000 == Info: connect to 127.0.0.1 port 9013 failed: Connection refused
        08:57:11.328000 == Info: Failed to connect to 127.0.0.1 port 9013: Connection refused
        08:57:11.328000 == Info: multi_done
        08:57:11.328000 == Info: Closing connection 0
        08:57:11.328000 == Info: The cache now contains 0 members
        === End of file trace2050

jay commented:
CI builds fail randomly due to timeouts sometimes the VM gets so slow it's inoperable. I've restarted the failed builds.

kunalekawde commented:
> 
> 
> CI builds fail randomly due to timeouts sometimes the VM gets so slow it's inoperable. I've restarted the failed builds.

Thanks, this time the builds are successful. :)

kunalekawde commented:
@jay / @bagder , please let me know if there is any action required from my side on this.

bagder commented:
@jay since you had remarks previously, I'll await your OK before I merge.  I have a local version prepared in a branch now with some additional minor edits.

bagder commented:
Thanks @kunalekawde for your hard work on this!

kunalekawde commented:
Thanks a lot @bagder , @jay for the guidance, learnt a lot during this first PR.
